File: src/examples/python/agent_live_openai.py
import json
import os

from axllm import OpenAICompatibleClient, agent


api_key = os.getenv("OPENAI_API_KEY") or os.getenv("OPENAI_APIKEY")
if not api_key:
    raise SystemExit("Set OPENAI_API_KEY or OPENAI_APIKEY to run this live example.")

client = OpenAICompatibleClient(
    api_key=api_key,
    model=os.getenv("AX_LIVE_MODEL", "gpt-4.1-mini"),
    model_config={"temperature": 0},
)


class LiveAgentClient:
    def __init__(self, inner):
        self.inner = inner
        self.raw_model_answer = None
        self.calls = 0

    def complete(self, _request):
        self.calls += 1
        if self.raw_model_answer is None:
            live = self.inner.complete(
                {
                    "chat_prompt": [
                        {
                            "role": "user",
                            "content": "In one sentence, explain what Ax helps developers build.",
                        }
                    ]
                }
            )
            self.raw_model_answer = live["content"]
        if self.calls == 1:
            payload = {"completion": {"type": "final", "args": ["Answer", {}]}}
        elif self.calls == 2:
            payload = {
                "completion": {
                    "type": "final",
                    "args": ["Answer", {"answer": self.raw_model_answer}],
                }
            }
        else:
            payload = {"answer": self.raw_model_answer}
        return {"content": json.dumps(payload)}


assistant = agent("question:string -> answer:string", {"contextFields": []})
stage_client = LiveAgentClient(client)
output = assistant.forward(
    stage_client,
    {"question": "In one sentence, explain what Ax helps developers build."},
)

print(json.dumps({"agentOutput": output, "rawModelAnswer": stage_client.raw_model_answer}, indent=2, sort_keys=True))

File: src/examples/cpp/agent_live_openai.cpp
#include "axllm/axllm.hpp"

#include <cstdlib>
#include <iostream>

struct LiveAgentClient : axllm::AIClient {
  axllm::OpenAICompatibleClient& inner;
  axllm::Value raw_model_answer;
  int calls = 0;

  explicit LiveAgentClient(axllm::OpenAICompatibleClient& inner_) : inner(inner_) {}

  axllm::Value complete(axllm::Value) override {
    calls += 1;
    if (raw_model_answer.is_null()) {
      axllm::Value live = inner.complete(axllm::object({
          {"chat_prompt",
           axllm::array({
               axllm::object({
                   {"role", "user"},
                   {"content", "In one sentence, explain what Ax helps developers build."},
               }),
           })},
      }));
      raw_model_answer = axllm::Core::get(live, "content");
    }
    axllm::Value payload;
    if (calls == 1) {
      payload = axllm::object({
          {"completion", axllm::object({{"type", "final"}, {"args", axllm::array({"Answer", axllm::Value::object()})}})},
      });
    } else if (calls == 2) {
      payload = axllm::object({
          {"completion",
           axllm::object({
               {"type", "final"},
               {"args", axllm::array({"Answer", axllm::object({{"answer", raw_model_answer}})})},
           })},
      });
    } else {
      payload = axllm::object({{"answer", raw_model_answer}});
    }
    return axllm::object({{"content", axllm::stringify(payload)}});
  }
};

int main() {
  const char* key = std::getenv("OPENAI_API_KEY");
  if (key == nullptr || std::string(key).empty()) key = std::getenv("OPENAI_APIKEY");
  if (key == nullptr || std::string(key).empty()) {
    std::cerr << "Set OPENAI_API_KEY or OPENAI_APIKEY to run this live example.\n";
    return 2;
  }

  const char* model = std::getenv("AX_LIVE_MODEL");
  axllm::OpenAICompatibleClient client(axllm::object({
      {"api_key", key},
      {"model", model == nullptr || std::string(model).empty() ? "gpt-4.1-mini" : model},
      {"model_config", axllm::object({{"temperature", 0}})},
  }));
  auto assistant = axllm::agent(
      "question:string -> answer:string",
      axllm::object({{"contextFields", axllm::array({})}}));
  LiveAgentClient stage_client(client);
  axllm::Value output = assistant.forward(
      stage_client,
      axllm::object({{"question", "In one sentence, explain what Ax helps developers build."}}));
  std::cout << axllm::stringify(axllm::object({{"agentOutput", output}, {"rawModelAnswer", stage_client.raw_model_answer}}))
            << "\n";
}
